firebase-firestore/src/main/java/com/google/firebase/firestore/remote/NetworkReachabilityMonitor.java

```diff
@@ -0,0 +1,32 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.remote;
+
+/** Interface for monitoring changes in network connectivity/reachability. */
+public interface NetworkReachabilityMonitor {
+  enum Reachability {
```

Comment:
To parallel iOS, this should be

```java
enum NetworkStatus {
    UNAVAILABLE,
    AVAILABLE,
    AVAILABLE_VIA_CELLULAR,  // if possible to determine
}
```

Reply:
This was deliberate. (Or rather, leaving off the cellular option was deliberate.) It's entirely possible to determine this (and I even had a preliminary implementation) but it's a bit messy, and requires two entirely separate implementations (one for N+, and one for M-). Since we don't actually use it or care about it, I left it off. Leaving it in ~ doubles the implementation. I've added a TODO. Alternatively, we can define the enum value, but never use it.

Re naming of the enum itself: done.

Reply:
iOS may not reference the enum value explicitly, but it still uses it here:
```cpp
void ConnectivityMonitor::MaybeInvokeCallbacks(NetworkStatus new_status) {
  if (new_status == status_) { // HERE
    return;
  }
  status_ = new_status;

  for (auto& callback : callbacks_) {
    callback(status_.value());
  }
}
```
If there wasn't a dedicated enum value for "cellular", the check `if (new_status == status_)` wouldn't be able to distinguish between the two "connected" states (Wi-Fi/cellular).

Android may not need the enum value, but it should also distinguish between Wi-Fi and cellular connections.

Reply:
It works a little different on android. We don't have a single network that changes status from wifi to cell; we have two different networks that (I suppose) could appear/disappear independently of each other. So a wifi->cell transition would look like the wifi network being lost, and the cell network becoming available.

That said, I'm going to go through another round of manual testing on this PR and I'll double check this particular comment to make sure I'm not confused.

Reply:
Cool, thanks for the detailed explanation.